core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala

```diff
@@ -1082,6 +1082,22 @@ class RDDSuite extends SparkFunSuite with SharedSparkContext {
     assert(totalPartitionCount == 10)
   }
 
+  test("SPARK-18406: race between end-of-task and completion iterator read lock release") {
+    val rdd = sc.parallelize(1 to 1000, 10)
+    rdd.cache()
+
+    rdd.mapPartitions { iter =>
+      ThreadUtils.runInNewThread("TestThread") {
```

Comment:
Nice use of this helper method. I wasn't aware of this, but it's pretty nice. I'll use it in my own tests going forward.